### ISSUE TITLE
[Feature Request] Add new way to manage timestep distribution and new loss offset feature

### ISSUE THREAD

**Issue by Anzhc** [kohya-ss/sd-scripts#1375]
We were working some time on an way to create timestep distribution specific to any particular dataset used in training.

We tested it for last couple months using Derrian easy scripts trainer with hardcoded changes, but we're not familiar with sd-scripts structure enough to fully integrate that as toggleable feature on our own at the current point in time, hence feature request.

https://github.com/Anzhc/Timestep-Attention-and-other-shenanigans

Here we give code of functions structured roughly correctly for usage with variables in sd-scripts. Also examples how we implemented it in hardcoded way are included in repo (look into "Kohya file examples" folder).
It builds distribution based on received loss values, similar to this:
![322636402-c5368e3c-c074-45a9-a929-e49808f87618](https://github.com/kohya-ss/sd-scripts/assets/133806049/a75b8150-360b-4ea0-919b-9bd934047c37)
But unique to each dataset(as it's based on particular loss values received in scope of training).

In majority of cases it improved convergence speed and quality of outcome in our tests.

We also provide code for Loss Offset Curve feature that works with said distribution, can be scaled and scheduled. It also supposed to improve convergence speed.

Some experimental loss functions are included too, but those are really just experiments.

**Comment by cheald** [comment]
This looks like it could pair really well with an approach that I'm working on and about to publish (the short version is "synchronoise noise std/mean to the underlying model"), which also dramatically improves convergence speed. I'm going to see if I can adapt this approach into my approach, because I have a theory that my approach will flatten the curves you're seeing there, which could potentially mean that the two approaches could cause faster convergence together.

Thanks!

**Comment by Anzhc** [comment]
> This looks like it could pair really well with an approach that I'm working on and about to publish (the short version is "synchronoise noise std/mean to the underlying model"), which also dramatically improves convergence speed. I'm going to see if I can adapt this approach into my approach, because I have a theory that my approach will flatten the curves you're seeing there, which could potentially mean that the two approaches could cause faster convergence together.
> 
> Thanks!

We actually were following your development on new loss approach and it's weighting for last couple days, would love to see that incorporated with your really fancy tech! :D

**Comment by cheald** [comment]
Can you elaborate a bit on the underlying theory of operation of your approach? A quick read through seemed to suggest that you're just tracking losses per-timestep, then weighting timestep selection towards higher-loss timesteps over the course of the run. Do I have that roughly correct?

**Comment by Anzhc** [comment]
Yes, that's the mechanism. We initialize a Loss Map that contains each timestep and it's loss, we initialize it with equal loss of 1 for each, but that can be changed to emulate different distributions at start, like lognorm(0,1).
Whenever we receive a new batch of timesteps, we update loss of that timestep and specified range of near timesteps by amount of weight decay, where last number is out main timestep, and each next is +-1 from it(i.e. if you put 5 numbers in it, it would change timestep received +-4 from it by specified portions). This helps to build map faster, but can be adjusted to be more precise for longer trainings.
It is also possible to vary averaging by increasing or decreasing change value, which is by default 60% move towards loss of timestep acquired.

Then we adjust that distribution by a small margin by dampening high chance timesteps and boosting low chance timesteps, to still give them meaningful chance to be sampled once in a while.

That's the mechanism, so you are correct.
Then that Timestep Loss Map can be used throughout training loop for weighting and adjusting various things, like loss curve function we have there, or clustered MSE, which selects cluster amount based on that map.

I have a gif how it changes in course of training,
![distribution_animation](https://github.com/kohya-ss/sd-scripts/assets/133806049/a8a4b552-8ca9-4650-9817-4a9b1309160a)

**Comment by Anzhc** [comment]
Different datasets led to meaningfully different graphs, so i think dynamic way of approaching this is currently best

**Comment by Anzhc** [comment]
![Screenshot 2024-06-19 025005](https://github.com/kohya-ss/sd-scripts/assets/133806049/4fd1a916-0430-4a47-8c82-187dc20bbb65)
![Screenshot 2024-06-19 025022](https://github.com/kohya-ss/sd-scripts/assets/133806049/1715f34e-4273-466b-87c7-e48329615e00)
Form is generally the same, though i think this is going to change on architecture-to-architecture basis, but spread in chances is roughly ~10-15% between those generally similar trainings(im usually just training some anime styles).

**Comment by cheald** [comment]
Are you using min_snr in your examples there? Based on my experiments, I would not expect that sharp downwards trend in loss observations in the earliest timesteps, unless you're using a mechanism that is artificially dampening it. Loss should be highest in the earliest timesteps (particular for SD1.5) as that's where the true noise/predicted noise standard deviation is the highest.

You might be able to simplify things by just using a gaussian kernel over the raw observations, which would give you configurable bandwidth for tuning the resolution/smoothness of the histogram.

**Comment by Anzhc** [comment]
I indeed do use min snr of 8 in training. I don't want to really drop it, but in hindsight it would be better to move `loss_for_timesteps` before it is altered by min snr and other things.
Right now im placing it after min snr and other mods, but before loss_curve_offset.

Im not sure simplifying would be the goal here, since i wasn't aiming at receiving smooth timestep distribution, but instead at a balance, where we still could see meaningful differences in sampling chance between relatively close timesteps, if they indeed reliably produce higher loss consistently.

When i was testing with some amount of Huber loss mixed in, trend was more even, just as a note.

I should place saving of loss for timesteps before min snr and see how it behaves with that.

**Comment by Anzhc** [comment]
My graphs were created with training with loss for timesteps being recorded at this point:
![image](https://github.com/kohya-ss/sd-scripts/assets/133806049/6f9857ef-2f9d-4024-98c8-67f0192053f1)

I'll put it before min snr and see what that leads me to whenever i have time to train some style again. Thanks for pointing that one out xd

**Comment by cheald** [comment]
Was this an SD1.5 model you generated the graphs on?

Absent the min_snr depression, this looks an awful lot like my observed std() distributions from my latest post in the loss thread, which makes me wonder if I could just use something like:

```python
prob = torch.nn.functional.softmax(-std_by_ts.mean(dim=1).reshape(-1), dim=0)
cat = torch.distributions.Categorical(probs=prob.float())
timesteps = cat.sample([b_size]).to(device=latents.device)
```

![image](https://github.com/kohya-ss/sd-scripts/assets/9830/2bfcd03a-7e4d-4170-ad5c-ac4d104a66eb)

![image](https://github.com/kohya-ss/sd-scripts/assets/9830/80587df6-5ada-4c43-a8bb-4dab9257c841)

From a quick first test, it seems to be VERY promising. I'll run some more comprehensive tests, but it might be that the std discrepancy is accounting for enough of the overall loss that we could just sample timestamps with it directly.

**Comment by Anzhc** [comment]
My graphs are from SDXL, Pony Diffusion v6 in particular.

It's hard for me to measure results by, your face, i think? Since i don't know it xd
On 1.5 i have very few tests, but dynamically sampled distribution did converge quite a bit better and faster when it came to particular concept i tested.
![323054378-decf0a08-de7f-493e-ac0b-60780b65a37d](https://github.com/kohya-ss/sd-scripts/assets/133806049/38d9c670-6c4d-4f52-93a9-3ffb63458ef4)
i.e. on epoch 15 it already looks better than result of standard distribution even on epoch 50.

It is probably will be quite beneficial to initialize our TA with your distribution to make it more efficient in early phase, and then let it modify itself to conform to dataset|model combo, as from your graphs i see that it does indeed change quite drastically based not only on dataset, but on underlying model as well as i thought.

I was thinking of doing general lognorm(0,1) initialization instead of uniform, or even shifted lognorm(0.5, 1) or so, that did show better performance in plain training on XL for us, but never tested that with dynamics.

Right now i started a training of style with `loss_for_timesteps` moved to before min snr, but it'll take some hours to run fully, and then i'll need to re-train another one with it moved back for reference.
Here is what distribution im at 220(1320 real steps) in:
![image](https://github.com/kohya-ss/sd-scripts/assets/133806049/c0e05318-0c1b-4b74-8f93-9423d1a8386d)
It doesn't seem like min snr affected distribution much in my case? This does look fairly similar to graphs i've shown before. Unless i f'ed something up, but i don't think that's the case?

**Comment by cheald** [comment]
Fair enough! Suffice it to say that both model quality and fidelity are way better than in previous approaches. 

I'll need to generate stats for ponyv6. It's a very different model than most and is probably going to exhibit some characteristics not found in other models. My sd1.5 result is conclusive enough that I think there's something very worth pursuing here!

**Comment by cheald** [comment]
Here's the noise std/mean for Ponyv6:

![image](https://github.com/kohya-ss/sd-scripts/assets/9830/d5b14e9f-618f-4211-afab-41f5340071f6)

As expected, it's quite different from SD1.5 or base SDXL models. The lack of mean convergence towards 0 is particularly interesting.

**Comment by bluvoll** [comment]
> Here's the noise std/mean for Ponyv6:
> 
> ![image](https://private-user-images.githubusercontent.com/9830/340914656-d5b14e9f-618f-4211-afab-41f5340071f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4NTcyMzgsIm5iZiI6MTcxODg1NjkzOCwicGF0aCI6Ii85ODMwLzM0MDkxNDY1Ni1kNWIxNGU5Zi02MThmLTQyMTEtYWZhYi00MWY1MzQwMDcxZjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjBUMDQxNTM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDYzNTFjMzAwNDU1MjBmNjViZmM5MjEzNTlhYWE5MTVkMzUwNjA5ZjkwOWU0YjNmYmYzYzg5MTJlMWY2YjNhOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.SeoG3HDIR-Bbec9vKma80BGNQDx7jyW_VS5nGlaYNnE)
> 
> As expected, it's quite different from SD1.5 or base SDXL models. The lack of mean convergence towards 0 is particularly interesting.

As you said, SDXL behaves very differently, in fact we got to the conclusion "SDXL is _very rigid_" even with extreme values it didn't reach unusable results like SD1.5, but we found that this approach still yielded better and faster results, in fact some stuff started to get more details when using lower LoRA Learning rate vs what we used to use with PonyV6, sadly we couldn't test  much Finetunning due to only having X090 class hardware, but results on rented A6000 and A100 showed the same tendecy of "being better", but we at lack a lot knowledge regarding ML in general, hopefully whatever you cook with this + your noise shenanigans ends up in something very very nice for the community.

**Comment by feffy380** [comment]
@Anzhc
> It doesn't seem like min snr affected distribution much in my case? This does look fairly similar to graphs i've shown before. Unless i f'ed something up, but i don't think that's the case?

I think it's the way you spread the loss weights. Values at the edges only receive updates from one side. Maybe use a 1d convolution with padding set to "replicate"

**Comment by Anzhc** [comment]
> I think it's the way you spread the loss weights. Values at the edges only receive updates from one side. Maybe use a 1d convolution with padding set to "replicate"

You mean let 999 be updated by 1 and vice versa? Might be an interesting idea. Tests from some people suggest that setting curve from the get go instead of letting it learn from uniform leads to worse performance in small trainings, so i think there is some grounding that needs to be done by learning earlier timesteps, and this likely will slightly increase hit at them.

**Comment by Anzhc** [comment]
But i wont be able to test anything for next 4 days e_e
![image](https://github.com/kohya-ss/sd-scripts/assets/133806049/3051cea6-3f7e-4b81-a3bd-b1a22c641eb1)
